### ISSUE TITLE
`no-constant-condition` should also count `typeof x`

### ISSUE THREAD

**Issue by isiahmeadows** [eslint/eslint#5228]
This, alone in any condition, is always truthy (always one of seven truthy values), and is probably a mistake in any case.

**Comment by isiahmeadows** [comment]
I think this could make it into 2.0 as well.

**Comment by eslintbot** [comment]
Hi @ilyavolodin, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by platinumazure** [comment]
Seems like this is an enhancement request. I would champion it, for what it's worth.

### Example 1

```js
if (typeof x) {
    // Constant condition
}
```

Current behavior: No warnings

Desired behavior: Should warn

### Example 2

```js
if (typeof x === "undefined") {
    // Non-constant condition
    // `typeof x` is constant in boolean terms but not constant in value
}
```

Current behavior: No warnings

Desired behavior: No warnings (this should join the valid tests if it isn't there already)

**Comment by platinumazure** [comment]
Oh, or alternatively, should this go under `valid-typeof`? The rule could be given an option to enforce that the typeof check is either assigned or compared via equality/inequality operator.

**Comment by mysticatea** [comment]
Interesting.

Either `no-constant-condition` or `valid-typeof` sounds reasonable to me.
I'll vote to `no-constant-condition`.
It's always truthy value for choice or loop conditions.

**Comment by ilyavolodin** [comment]
Agree. While this is a constant condition that never changes, it's also a valid typeof, so I don't think it's a good fit. `no-constant-condition` enhancement sounds good.

**Comment by nzakas** [comment]
Not sure I can get behind this change. We have now crossed over from simple static analysis to interpreting the meaning of an expression. What we flag right now are obvious boolean conditions, but this is coercion, and more likely than not a bug, do we really want to start adding these types of conditions?

**Comment by platinumazure** [comment]
@nzakas What about a standalone rule for this (UnaryExpression with typeof, not in a comparing BinaryExpression, in boolean context such as IfStatement condition)? Or as an extra augmentation of `valid-typeof` per my suggestion above?

**Comment by isiahmeadows** [comment]
I would be fine with a new rule for those suggestions (there's a few other
cases where `typeof` could be pointless, such as checking `typeof x ===
"undefined"` when `x` is already defined within that scope, but those
should get their own rule).

But I think this specific suggestion of mine should just augment an
existing rule. Unlike some of the other suggestions, my initial request
hardly adds anything. It marks a very likely bug (you might as well use the
`void` operator), and the only context you need to check for this is parent
node's type.

**Comment by michaelficarra** [comment]
@isiahmeadows `typeof` can produce a falsy value through use of `Symbol.toStringTag`. Though for the purposes of a linter, it may be okay to ignore that.

**Comment by nzakas** [comment]
I don't think this rises to the level of a standalone core rule. I still haven't seen a good argument as to why this should be included in `no-constant-condition`. @ilyavolodin @mysticatea can either of you explain your rationale, since it seemed like you were in favor of this change?

**Comment by ilyavolodin** [comment]
```js
if (typeof variable) {}
```
is a constant condition. It's most likely a mistake and easy enough to catch. I think it's a reasonable addition that doesn't add much complexity. I'm not sure I see a huge different between the example above and ``if (true) {}```.

**Comment by nzakas** [comment]
Hmmm OK, I'm not completely sold, but if you feel strongly, go for it.

**Comment by isiahmeadows** [comment]
@Michael How is `typeof` affected by Symbol.toStringTag? Last time I
checked, that's only Object.prototype.toString that's affected. And also,
an empty string results in "[object ]" in that case.

**Comment by michaelficarra** [comment]
@isiahmeadows Ah, sorry. I meant new user-defined value types, which are not available yet. I don't know how `Symbol.toStringTag` came out. Must have been tired.

**Comment by isiahmeadows** [comment]
Okay. In that light, it still should be an option that can toggled, since until then, my initial assertion still holds, unless [the implementation defines its own type that return something different](http://www.ecma-international.org/ecma-262/6.0/#sec-typeof-operator) (which is very strongly discouraged by the spec).

**Comment by nzakas** [comment]
@isiahmeadows do you want to submit a PR?

**Comment by isiahmeadows** [comment]
When I get time, I may. I have never messed with custom rules before,
though, much less ones in ESLint proper.

**Comment by pvamshi** [comment]
am working on this

**Comment by michaelficarra** [comment]
If `typeof` in boolean position is reported by this rule, `void` should be as well.

**Comment by nzakas** [comment]
@michaelficarra can you open a separate issue for that?

**Comment by pvamshi** [comment]
If you are creating different task, I will ignore void and take care of typeof alone in this issue. Can you guys please confirm.

**Comment by michaelficarra** [comment]
I don't see why it's different. They're both unary operators that return an unconditionally truthy/falsey value.

**Comment by platinumazure** [comment]
@michaelficarra It's just about scope creep for the issue, I think. If we
want void accounted for as well, I have no problem with that- but it should
be tracked separately since the original issue was about typeof.

**Comment by nzakas** [comment]
@michaelficarra what @platinumazure said. The PR was already in progress, and it's unfair to contributors to move the finish line once they've started on a solution. In general, any comment that's "and also" to an issue should end up as its own issue for tracking purposes.
